python/ray/tune/automl/search_policy.py

```diff
@@ -0,0 +1,237 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import time
+import copy
+import logging
+
+from ray.tune.trial import Trial
+from ray.tune.suggest import SearchAlgorithm
+from ray.tune.experiment import convert_to_experiment_list
+from ray.tune.suggest.variant_generator import generate_variants
+from ray.tune.config_parser import make_parser, create_trial_from_spec
+
+logger = logging.getLogger(__name__)
+
+
+def deep_insert(path_list, value, config):
+    """Inserts value into config by path, generating intermediate dictionaries.
+
+    Example:
+        >>> deep_insert(path.split("."), value, {})
+    """
+    if len(path_list) > 1:
+        inside_config = config.setdefault(path_list[0], {})
+        deep_insert(path_list[1:], value, inside_config)
+    else:
+       config[path_list[0]] = value
+
+
+class AutoMLSearcher(SearchAlgorithm):
+    """Base class for AutoML search algorithm
+
+    It works in a round-by-round way. For each experiment round,
+    it generates a bunch of parameter config permutations, submits
+    and keeps track of them. Once all of them finish, results will
+    be fed back to the algorithm as a whole.
+    """
+
+    CONTINUE = "CONTINUE"
+    TERMINATE = "TERMINATE"
+
+    def __init__(self, search_space, reward_attr):
+        """Initialize AutoMLSearcher.
+
+        Arguments:
+            search_space (SearchSpace): The space to search.
+            reward_attr: The attribute name of the reward in the result.
+        """
+        # Pass experiment later to allow construction without this parameter
+        super(AutoMLSearcher, self).__init__()
+
+        self.search_space = search_space
+        self.reward_attr = reward_attr
+
+        self.experiment_list = []
+        self.best_trial = None
+        self._is_finished = False
+        self._parser = make_parser()
+        self._undone_count = 0
+        self._running_trials = {}
+        self._completed_trials = {}
+
+        self._iteration = 0
+        self._total_trial_num = 0
+        self._start_ts = 0
+
+    def add_configurations(self, experiments):
+        self.experiment_list = convert_to_experiment_list(experiments)
+
+    def get_best_trial(self):
+        """Returns the Trial object with the best reward_attr"""
+        return self.best_trial
+
+    def next_trials(self):
+        if self._undone_count > 0:
+            # Last round not finished
+            return []
+
+        trials = []
+        raw_param_list, extra_arg_list = self._select()
```

Comment:
nit: perhaps `extra_arg_list` should just return the None list, instead of handling it here?